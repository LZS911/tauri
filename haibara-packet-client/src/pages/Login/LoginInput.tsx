import { EyeOutlined, EyeInvisibleOutlined } from '@ant-design/icons';
import { useBoolean, useToggle } from 'ahooks';
import { forwardRef, HTMLInputTypeAttribute, useMemo } from 'react';
import { LoginInputProps } from '.';
import ThemeBase from '../../components/ThemeBase';
import usePropsValue from '../../hooks/usePropsValue';

const LoginInput: React.ForwardRefRenderFunction<
  HTMLDivElement,
  LoginInputProps
> = (
  {
    placeholder,
    prefix,
    value,
    onChange,
    defaultValue = '',
    label,
    type = 'text',
    name,
  },
  ref
) => {
  const [internalValue, setInternalValue] = usePropsValue<string>({
    defaultValue,
    value,
    onChange,
  });
  const [isFocus, { setFalse: clearFocus, setTrue: setFocus }] =
    useBoolean(false);
  const [isShowPassword, { toggle: togglePasswordVisibility }] =
    useToggle(false);
  const internalType = useMemo((): HTMLInputTypeAttribute => {
    if (type !== 'password') {
      return type;
    }

    return isShowPassword ? 'text' : 'password';
  }, [isShowPassword, type]);
  return (
    <>
      <div ref={ref} className="mb-1 text-xs">
        <span className="text-xs text-black opacity-90 dark:text-white">
          {label}
        </span>
      </div>
      <div
        className={`flex w-full items-center border-b-2 pb-2 pl-2 ${
          isFocus ? 'border-primary' : 'border-gray-300'
        } transition-colors`}
      >
        <span className="dark:text-white">{prefix}</span>
        <input
          id={name}
          type={internalType}
          className={`ml-3 h-full w-4/5 p-2 outline-0 dark:bg-darkMode dark:text-white`}
          value={internalValue}
          onChange={(e) => {
            setInternalValue(e.target.value);
          }}
          placeholder={placeholder}
          onFocus={setFocus}
          onBlur={clearFocus}
        />
        {type === 'password' && (
          <ThemeBase.Icon
            onClick={togglePasswordVisibility}
            icon={isShowPassword ? <EyeOutlined /> : <EyeInvisibleOutlined />}
          />
        )}
      </div>
    </>
  );
};

export default forwardRef(LoginInput);
