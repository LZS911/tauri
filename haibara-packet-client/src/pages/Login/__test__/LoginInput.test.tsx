import { fireEvent, render, screen } from '@testing-library/react';
import { renderWithRedux } from '../../../testUtils/customRender';
import LoginInput from '../LoginInput';

const placeholder = 'this is placeholder';
const prefix = 'prefix';
const label = 'label';
const defaultValue = 'defaultValue';
const value = 'value';
const onChange = jest.fn();

describe('test LoginInput', () => {
  afterEach(() => {
    jest.clearAllTimers();
    jest.clearAllMocks();
  });
  test('should match snapshot', () => {
    const { container } = render(
      <LoginInput placeholder={placeholder} prefix={prefix} label={label} />
    );

    expect(container).toMatchSnapshot();

    fireEvent.focus(container.querySelector('input')!);
    expect(container).toMatchSnapshot();
  });

  test('should set defaultValue when defaultValue is defined', () => {
    const { container } = render(
      <LoginInput
        placeholder={placeholder}
        prefix={prefix}
        label={label}
        defaultValue={defaultValue}
      />
    );

    expect(container.querySelector('input')).toHaveValue(defaultValue);
  });

  test('should be able to switch types when type is password', () => {
    const { container } = renderWithRedux(
      <LoginInput
        placeholder={placeholder}
        prefix={prefix}
        label={label}
        type="password"
      />
    );

    expect(container.querySelector('input')).toHaveAttribute(
      'type',
      'password'
    );
    expect(screen.getByRole('img')).not.toBeUndefined();
    fireEvent.click(screen.getByRole('img'));
    expect(container.querySelector('input')).toHaveAttribute('type', 'text');
  });

  test('should trigger onChange props when modifying input value', () => {
    const { container } = render(
      <LoginInput
        placeholder={placeholder}
        prefix={prefix}
        label={label}
        value={value}
        onChange={onChange}
      />
    );

    expect(onChange).toBeCalledTimes(0);
    fireEvent.change(container.querySelector('input')!, {
      target: { value: 'test' },
    });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith('test');
  });
});
