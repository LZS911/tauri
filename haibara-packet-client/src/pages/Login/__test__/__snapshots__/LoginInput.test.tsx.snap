// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test LoginInput should match snapshot 1`] = `
<div>
  <div
    class="mb-1 text-xs"
  >
    <span
      class="text-xs text-black opacity-90 dark:text-white"
    >
      label
    </span>
  </div>
  <div
    class="flex w-full items-center border-b-2 pb-2 pl-2 border-gray-300 transition-colors"
  >
    <span
      class="dark:text-white"
    >
      prefix
    </span>
    <input
      class="ml-3 h-full w-4/5 p-2 outline-0 dark:bg-darkMode dark:text-white"
      placeholder="this is placeholder"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`test LoginInput should match snapshot 2`] = `
<div>
  <div
    class="mb-1 text-xs"
  >
    <span
      class="text-xs text-black opacity-90 dark:text-white"
    >
      label
    </span>
  </div>
  <div
    class="flex w-full items-center border-b-2 pb-2 pl-2 border-primary transition-colors"
  >
    <span
      class="dark:text-white"
    >
      prefix
    </span>
    <input
      class="ml-3 h-full w-4/5 p-2 outline-0 dark:bg-darkMode dark:text-white"
      placeholder="this is placeholder"
      type="text"
      value=""
    />
  </div>
</div>
`;
